### ISSUE TITLE
localcomapre with empty string crashes hermes in android app

### ISSUE THREAD

**Issue by punksta** [facebook/hermes#602]
## Bug Description

```
function crashMe() {
  '1'.localeCompare('')
}
```

- [x ] I have run `gradle clean` and confirmed this bug does not occur with JSC

ios and jsc are working fine.

Hermes version: 0.8.1
React Native version (if any): 0.65.1
OS version (if any): 29, 31
Platform: arm64-v8a, x86_64, (most likely one of arm64-v8a, armeabi-v7a, x86, x86_64):

## Steps To Reproduce

invoke localCompare with empty string.

```
A/libc: /buildbot/src/googleplex-android/ndk-release-r20/external/libcxx/../../external/libcxxabi/src/abort_message.cpp:73: abort_message: assertion "terminating with uncaught exception of type facebook::jni::JniException: java.lang.NullPointerException: Attempt to invoke interface method 'int java.lang.CharSequence.length()' on a null object reference" failed
```

code example:
```
 '1'.localeCompare('')
```

**Comment by SudoPlz** [comment]
Are there any work arounds until this get's implemented?

There are various libraries that we're using that heavily rely on `localeCompare`, and our app is crashing at launch time.
Has anyone found any way to work around that issue?

At first I thought, well ok, I can swizzle `localeCompare` with `Intl.Collator` but that isn't supported either (!!) by hermes.
Is the only way to keep the app up and running to stop using hermes for now?

Thanks

**Comment by punksta** [comment]
```
function patchLocalCompare() {
  const originalLocalCompare = String.prototype.localeCompare

  const compareStrings = (a, b) => {
    if (a === b) {
      return 0
    } else if (a === '') {
      return -1
    } else if (b === '') {
      return 1
    } else {
      return originalLocalCompare.call(a, b)
    }
  }

  String.prototype.localeCompare = (that) => {
    return compareStrings(this, that)
  }
}

patchLocalCompare();
 ```
 
You can try to monkey patch localeCompare as a work around. Try adding this to index.js before other libraries.

**Comment by SudoPlz** [comment]
@punksta ok this worked like a charm.

I'm using the following code:

```
export default function swizzleLocaleCompare() {
  const originalLocalCompare = String.prototype.localeCompare;
  const compareStrings = (a: string, b: string, locales?: string | string[], options?: Intl.CollatorOptions) => {
    if (a === b) {
      return 0
    } else if (a === '') {
      return -1
    } else if (b === '') {
      return 1
    } else {
      return originalLocalCompare.call(a, b, locales, options)
    }
  }

  String.prototype.localeCompare = (that: string, locales?: string | string[], options?: Intl.CollatorOptions) => {
    // @ts-ignore
    return compareStrings(this, that, locales, options);
  }
}
```

(to make sure locales, and options are also supported)

**Comment by neildhar** [comment]
@punksta thank you for reporting this, it looks like a bug in fbjni's `make_jstring` returning `Null` for empty strings. I'll put up a fix.

**Comment by neildhar** [comment]
Fixed in ce59ce1

**Comment by igium** [comment]
In case somebody still uses an older version showing this problem, this is a piece of code that works form me:

```javascript
(function()
{
    const String_localeCompare = String.prototype.localeCompare;
    String.prototype.localeCompare = function localeCompare(b, ...args)
    {
        try
        {
            if (this === b) return 0;
            else if (this === "") return -1;
            else if (b === "") return 1;
            else return String_localeCompare.call(this, b, ...args);
        }
        catch (ex)
        {
            console.error(666, ex);
        }
    }
})();
```

**Comment by gnprice** [comment]
As a breadcrumb for anyone else running into this in a React Native app:

The fix above went out in Hermes v0.11.0. The RN version that in turn took that Hermes update was v0.68.0.
